### PR TITLE
generic:sycl: Inner Product FWD

### PR DESCRIPTION
**Pull request by sgeor255**
# Checklist

## General

- [X] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
- [X] Have you formatted the code using clang-format?

### CODE REVIEW

**Review by Rbiessy — APPROVED** [review]

File: src/gpu/generic/sycl/ref_inner_product.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+status_t ref_inner_product_fwd_t::pd_t::init_matmul(impl::engine_t *engine) {
+    matmul_desc_t matmul_desc;
+    CHECK(matmul_desc_init(&matmul_desc, &src_md_reshaped, &weights_md_reshaped,
+            &bias_md_reshaped, arg_md(DNNL_ARG_DST)));
+    primitive_attr_t matmul_attr(*attr());
+
+    primitive_desc_iterator_t it(engine,
+            reinterpret_cast<op_desc_t *>(&matmul_desc), &matmul_attr, nullptr);
+    if (!it.is_initialized()) return status::invalid_arguments;
```
Comment (ShanoToni) [inline src/gpu/generic/sycl/ref_inner_product.cpp:31]: Seems the correct status to be returned here is `status::out_of_memory`

**Review by ShanoToni — APPROVED** [review]

File: src/gpu/nvidia/cudnn_matmul_executor.hpp [hunk 2]
```diff
@@ -392,12 +392,12 @@ struct cudnn_matmul_lt_exec_t final : public cudnn_matmul_lt_base_exec_t {
                             memory_tracking::names::key_matmul_dst_in_acc_dt)
                     : xpu::sycl::interop_memory_arg_t<
                             ::sycl::access::mode::read_write>();
-            auto arg_block_a_scratch = params->source_size_ != 0
+            auto arg_block_a_scratch = params->weight_size_ != 0
```
Comment (mgouicem) [inline src/gpu/nvidia/cudnn_matmul_executor.hpp:395]: (random spot) the cudnn_matmul commits seem to be unrelated to generic inner product.
Could you commit those fixes in a separate commit (same PR is fine), this way if one or the other needs to be reverted, we don't revert everything.

**Review by mgouicem — APPROVED** [review]

File: src/gpu/nvidia/cudnn_matmul_executor.hpp (hunk 2 above)
Comment (Rbiessy) [reply src/gpu/nvidia/cudnn_matmul_executor.hpp:395]: Sure, this is done

**Review by vpirogov — APPROVED** [review]

File: src/gpu/generic/sycl/ref_inner_product.hpp [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            const auto weights_md_ = arg_md(DNNL_ARG_WEIGHTS);
+            const auto bias_md_ = arg_md(DNNL_ARG_BIAS);
+            auto src_wrap = memory_desc_wrapper(src_md_);
+            auto w_wrap = memory_desc_wrapper(weights_md_);
+
+            // src and weights dims need to be in the same order
+            if (get_dim_order(src_wrap.ndims(), src_wrap.strides())
+                    != get_dim_order(w_wrap.ndims(), w_wrap.strides())) {
+                return status::unimplemented;
+            }
```
Comment (Rbiessy) [inline src/gpu/generic/sycl/ref_inner_product.hpp:127]: Highlighting this as we are currently skipping the cases where src and weights are in a different order. We can consider supporting this in the future using reorder.
